src/EventStore.Core/Services/Archive/Archiver/ArchiverService.cs

```diff
@@ -3,240 +3,116 @@
-	IHandle<SystemMessage.ChunkLoaded>,
-	IHandle<SystemMessage.ChunkCompleted>,
+public sealed class ArchiverService :
+	IHandle<SystemMessage.SystemStart>,
+	IHandle<SystemMessage.BecomeShuttingDown>,
 	IHandle<SystemMessage.ChunkSwitched>,
 	IHandle<ReplicationTrackingMessage.ReplicatedTo>,
-	IHandle<SystemMessage.BecomeShuttingDown>
-{
-	private static readonly ILogger Log = Serilog.Log.ForContext<ArchiverService>();
+	IAsyncDisposable {
 
-	private readonly ISubscriber _mainBus;
 	private readonly IArchiveStorage _archive;
-	private readonly Queue<ChunkInfo> _uncommittedChunks;
-	private readonly ConcurrentDictionary<string, ChunkInfo> _existingChunks;
-	private readonly CancellationTokenSource _cts;
-	private readonly Channel<Commands.ArchiveChunk> _archiveChunkCommands;
-	private readonly IChunkUnmerger _chunkUnmerger;
+	private readonly CancellationToken _lifetimeToken;
+	private readonly AsyncAutoResetEvent _archivingSignal;
+	private readonly IChunkRegistry<IChunkBlob> _chunkManager;
+	private readonly ConcurrentBag<ChunkInfo> _switchedChunks;
+	private Task _archivingTask;
 
-	private readonly TimeSpan RetryInterval = TimeSpan.FromMinutes(1);
-	private long _replicationPosition;
-	private bool _archivingStarted;
-	private long _checkpoint;
+	private long _replicationPosition; // volatile
+	private volatile CancellationTokenSource _cts;
 
+	// systemStart
 	public ArchiverService(
 		ISubscriber mainBus,
 		IArchiveStorage archiveStorage,
-		IChunkUnmerger chunkUnmerger) {
-		_mainBus = mainBus;
+		IChunkRegistry<IChunkBlob> chunkChunkManager) {
 		_archive = archiveStorage;
-		_chunkUnmerger = chunkUnmerger;
-
-		_uncommittedChunks = new();
-		_existingChunks = new();
 		_cts = new();
-		_archiveChunkCommands = Channel.CreateUnboundedPrioritized(
-			new UnboundedPrioritizedChannelOptions<Commands.ArchiveChunk> {
-				SingleWriter = false,
-				SingleReader = true,
-				Comparer = new ChunkPrioritizer()
-			});
-
-		Subscribe();
-	}
-
-	private void Subscribe() {
-		_mainBus.Subscribe<SystemMessage.ChunkLoaded>(this);
-		_mainBus.Subscribe<SystemMessage.ChunkSwitched>(this);
-		_mainBus.Subscribe<SystemMessage.ChunkCompleted>(this);
-		_mainBus.Subscribe<ReplicationTrackingMessage.ReplicatedTo>(this);
-		_mainBus.Subscribe<SystemMessage.BecomeShuttingDown>(this);
+		_lifetimeToken = _cts.Token;
+		_archivingSignal = new(initialState: false);
+		_chunkManager = chunkChunkManager;
+		_archivingTask = Task.CompletedTask;
+		_switchedChunks = [];
+
+		mainBus.Subscribe<SystemMessage.ChunkSwitched>(this);
+		mainBus.Subscribe<ReplicationTrackingMessage.ReplicatedTo>(this);
+		mainBus.Subscribe<SystemMessage.BecomeShuttingDown>(this);
 	}
 
-	public void Handle(SystemMessage.ChunkLoaded message) {
-		if (!message.ChunkInfo.IsCompleted)
-			return;
-
-		// queue chunk for archiving
-		_existingChunks[Path.GetFileName(message.ChunkInfo.ChunkLocator)!] = message.ChunkInfo;
+	public void Handle(SystemMessage.SystemStart message) {
```

Comment:
the archiver node may need to truncate when it joins the leader, so it might not be safe to start archiving at this point

Reply:
it's ok because its guarded by the replication checkpoint though?

Reply:
I'm not sure, but the truncator runs at very early stage of the node, in `ClusterVNode.StartNode` method. According to the source code, `TruncateDb()` happens before `_mainQueue.Start();`. I can assume that `SystemStart` happens after `TruncateDb()`.

Reply:
it's a different part of the truncation that shaan is concerned about, when it makes the decision to truncate rather the actual truncation.

still, i think its ok because we're checking the replication checkpoint, unless @shaan1337 sees it differently

Reply:
i think that it's fine right now, as the replication checkpoint is initialized to -1. but say in the future, somebody comes along and changes the replication checkpoint to be initialized to another value (e.g the index checkpoint, which is also a valid initialization value), then the behaviour could be wrong:

e.g consider the case where the (leader's and follower's) data is restored from a backup but the archiver is still ahead of them: the archiver would already start uploading chunks to the archive before it joins the cluster, then it would truncate.

i'm not sure if we need to worry about that case for now though.

Reply:
yeah i dont think we need to worry about that, it makes sense to me for the archiver to be allowed to archive anything that it knows is replicated. if we want to stop it from doing that because we've restored a backup to multiple nodes, then we ought to restore that backup to the archiver node as well